src/base-app.js

```diff
@@ -32,59 +33,65 @@ class FusionApp {
     this.register(createPlugin({deps, middleware}));
   }
   enhance(token, enhancer) {
-    const {value, aliases, enhancers} = this.registered.get(token) || {
+    const {value, aliases, enhancers} = this.registered.get(
+      token.ref || token
+    ) || {
       aliases: new Map(),
       enhancers: [],
     };
     enhancers.push(enhancer);
-    this.registered.set(token, {value, aliases, enhancers});
+    this.registered.set(token.ref, {value, aliases, enhancers});
   }
   resolve() {
-    const resolved = new Map();
-    const dependedOn = new Set();
-    const nonPluginTokens = new Set();
-    const resolving = new Set();
-    const registered = this.registered;
-    const resolvedPlugins = [];
-    const allAliases = new Set();
+    const resolved = new Map(); // Token.ref || Token => Service
+    const dependedOn = new Set(); // Token.ref || Token
+    const nonPluginTokens = new Set(); // Token
+    const resolving = new Set(); // Token.ref || Token
+    const registered = this.registered; // Token.ref || Token -> {value, aliases, enhancers}
+    const resolvedPlugins = []; // Plugins
+    const allAliases = new Set(); // Token.ref || Token
     const resolveToken = (token, tokenAliases) => {
```

Comment:
Unfortunately `token` here can be either a `Token` (new type) or any arbitrary value (which likely includes `Plugins`).  As such, we have to check truthy-ness of `token.ref` and `token` which isn't particularly elegant.  Open to suggestions on how to clean this up since this code is fairly complex already.

Reply:
I'd argue for having `createToken` in this package, and doing an `instanceof` check

Reply:
+1, we could do something like:

```js
// Helpers
function getTokenRef(token: TokenImpl | any): any {
   if(token instanceof TokenImpl) {
      return token.ref;
   }
   return token;
}

And replace each `token.ref || token` with `getTokenRef(token)`. 
```